Docvelkv

Docvelkv opened the pull request.

Commit:
Вроде написал, требует осмысления и доработки.

Docvelkv committed:
Переделал FamilyTree, Human, Main.

Liberate520 commented on src/familyTree/Human.java:
```diff
@@ -0,0 +1,210 @@
+    private Gender gender;
+    private LocalDate dateOfBirth;
+    private LocalDate dateOfDeath;
+    private String mother;
+    private String father;
+    private List<String> spouse;
+    private List<String> childs;
```
Лучше все таки хранить ссылку не на String, а на Human. Human mother

Liberate520 commented on src/familyTree/FamilyTree.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Демонстрация всех членов древа
+     */
+    public void showAll() {
+        int id = 1;
+        for (Map.Entry<String, Human> human: wholeGenus.entrySet()) {
+            System.out.printf("%d. %s\n__________\n", id, human.getValue());
+            id++;
+        }
+    }
+
+    public void showHuman(String partFullName) {
+        int res = 0;
+        for (Map.Entry<String, Human> human: wholeGenus.entrySet()) {
+            if (human.getKey().contains(partFullName)){
+                System.out.printf("%s\n__________\n", human.getValue());
+                res ++;
+            }
+        }
+        if (res == 0) System.out.println("Такого человека не найдено.");
+    }
+}
```
лучше чтобы методы также как и в классе Human возвращали текст, а не печатали его в консоль

Docvelkv committed:
Исправил методы FamilyTree

Docvelkv committed:
Исправил методы FamilyTree и класс Human

Docvelkv committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/familyTree/FamilyTree.java

Docvelkv committed:
Сериализация

Docvelkv committed:
Сериализация. Создал метод для сохранения.

Docvelkv committed:
Сериализация. Всё переделал.
Создал отдельные класс и интерфейс для сохранения. Все классы разложил по пакетам.

Docvelkv committed:
Исправил работу методов при значении null. Переписал toString. Сделал мелкие исправления.

Docvelkv committed:
Сделал класс для хранения нескольких FamilyTree. Сделал сортировку FamilyTree по ключу.

Liberate520 commented on src/familyTree/tree/FamilyTree.java:
```diff
@@ -0,0 +1,87 @@
+     * Поэтому сделал только так.
+     */
+    public void sortByAge(){
+        List<Integer> list = new ArrayList<>();
+        Map<String, Human> sortedMap = new LinkedHashMap<>();
+        for (Map.Entry<String, Human> human : this.wholeGenus.entrySet())
+            list.add(human.getValue().getAge());
+        Collections.sort(list);
+        for (int i : list) {
+            for (Map.Entry<String, Human> human : this.wholeGenus.entrySet())
+                if (human.getValue().getAge() == i){
+                    sortedMap.put(human.getKey(), human.getValue());
+                }
+        }
+        wholeGenus = sortedMap;
+        System.out.println(this.showAll());
```
А класс компаратора тогда зачем? Здесь должно было быть Map<String, Human> map = new TreeMap(new HumanComparatorByAge); и далее перекладываем все элементы и меняем ссылку wholeGenus

Docvelkv committed:
Ура понял!!! Сделал второй вариант сортировки по возрасту со своим компаратором.

Docvelkv committed:
Дженерики + некоторые небольшие изменения.

Liberate520 commented on src/familyTree/tree/FamilyTree.java:
```diff
@@ -0,0 +1,118 @@
+     * @param fullName полное имя
+     * @return объект типа Human
+     */
+    public Human searchHumanByFullName(String fullName){
+        return (Human) wholeGenus.getOrDefault(fullName, (T) new Human(""));
+    }
```
Ничего не получится. А если интерфейс FamilyTreeNode будет реализовывать класс отличный от Human? Ваше преобразование типов уронит приложение. Данный метод может быть реализован в сервисе. Если человек не нашелся, то именно сервис может создать человека (Сервис должен быть написан под конкретный тип данный, под человека в нашем случае)

Docvelkv committed:
Переделал под MVP + другие мелкие изменения.

Liberate520 commented on src/familyTree/model/ServiceHuman.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class ServiceHuman implements Serializable {
```
лишний интерфейс. Данный интерфейс используется для объектов, которые мы собираемся сохранять, но объект сервиса мы не собираемся сохранять

Liberate520 commented on src/familyTree/model/ServiceHuman.java:
```diff
@@ -0,0 +1,189 @@
+
+    private int id;
+    private DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+    private final Map<String, FamilyTree> TREE_GROUP = new LinkedHashMap<>();
```
все таки это не совсем константа в нашем понимании. ссылка не меняется, это да, но значения же изменяются. Лучше капслок здесь не применять

Docvelkv committed:
Всё переделал под десктопное приложение

Liberate520 commented on src/main/java/oop/familyTree/MVP/model/ServiceHuman.java:
```diff
@@ -0,0 +1,135 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ServiceHuman implements SaveAndLoad{
```
Неверный интерфейс. Данный интерфейс ознакчает классы, которые по идее можно использовать для сохранения и загрузки, но у класса сервис другие задачи стоят. То что у него есть методы сохранения и загрузки, еще не означает, что ему нужно подписаться на этот интерфейс

Liberate520 commented on src/main/java/oop/familyTree/MVP/model/ServiceHuman.java:
```diff
@@ -0,0 +1,135 @@
+package oop.familyTree.MVP.model;
+
+import javafx.scene.control.Alert;
```
Зависимость модели от классов View это ошибка. Сообщение нужно возвращать как текст и уже View должно его отобразить так, как считает нужным

Liberate520 commented on src/main/java/oop/familyTree/MVP/model/ServiceHuman.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    public FamilyTree<Human> loading(String treeName) {
+        Path dirPath = Paths.get(pathDirFamTree + "/" + treeName + ".ser");
+        try(ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(dirPath))) {
+            return (FamilyTree<Human>) ois.readObject();
+        }
```
Должен быть отдельный класс реализующий задачи работы с файловой системой. Сервис уже посредством этого класса выполняет сохранение и загрузку

Docvelkv committed:
Консольное приложение готово

Docvelkv committed:
Переписал "Изменение значения поля" (шаблон фабрика)

Docvelkv committed:
Мелкие изменения

Liberate520 commented on src/main/java/oop/familyTree/mvp/presenter/Presenter.java:
```diff
@@ -0,0 +1,117 @@
+     * Конструктор
+     */
+    public Presenter() {
+        this.service = new Service();
+        new DesktopController();
```
лучше два отдельных презентера было создать, чтобы не создавать лишних сущностей
